\documentclass{article}
\usepackage{graphicx}
\usepackage{amsmath,amssymb,amsthm}
\usepackage{color}
\usepackage{fullpage}

\def\xcolorversion{2.00}
\def\xkeyvalversion{1.8}

\usepackage[version=0.96]{pgf}
\usepackage{tikz}
\usetikzlibrary{arrows, backgrounds, positioning, fit, shapes, automata}
\usepackage[latin1]{inputenc}

\usepackage{algpseudocode}
\usepackage{algorithm}

\DeclareMathOperator*{\argmax}{arg\,max}
\DeclareMathOperator*{\argmin}{arg\,min}

\begin{document}

\title{De Novo Variant Caller using Google Genomics API}

\author{
	Subhodeep Moitra\\ 
	Carnegie Mellon University \\
	{\tt subhodee@cs.cmu.edu}
}

\maketitle

\begin{abstract}
This tool identifies de novo genetic mutations in a family trio (child, mother and father). De novo genetic mutations are defined as those mutations which occur in the child but do not occur in the genomes of the parents. The tool queries the genomics API in order to search and retrieve genetic variants and reads. It implements a Bayesian inference algorithm for identifying  de novo mutations from the retrieved data. The tool also provides options for customizing the inference algorithm and the search region. This tool can be used by customers of google genomics for their analysis needs and also provides an example of code that interacts with the Genomics API for a non-trivial scientific analysis task.
\end{abstract} 

\section{Background}

\subsection{Bayes Net}
The de novo variant caller is implemented using a bayes net. A bayes net belongs to the class of mathematical models known as graphical models which are a marriage between graph theory and probability theory~\cite{Jordan2008}. A Bayes Net is a directed acyclic graph which defines conditional independence relationships between random variables in a probability distribution. A Bayes net can be used for sampling, marginal inference and decoding of the hidden variables. 

\vspace{5mm} 
We follow the bayes net implementation for de novo variant calling as in Li et al~\cite{Li2012}. They model a bayes net to encode the diploid Mendelian relationship between the members of the trio. See Figure~\ref{fig:bayesnet} for an illustration of the bayes net. The hidden variables $G_{F}, G_{M} and G_{C}$ correspond to the genotypes of the father, mother and child respectively. The genotypes take one among 10 values from $AA, AC, AT, AG, CC, CT, CG, GG, GT$ and $TT$. The trio of genotypes values can thus have $10^3$ possible values. The observed variables ($D_F, D_M$ and $D_C$) are the bases present at the candidate position for a particular read. 

\begin{equation}\label{eqn:model}
P(G_F,G_M,G_C|D_F,D_M,D_C) =  \frac{P(D_F,D_M,D_C|G_F,G_M,G_C)P(G_F,G_M,G_C)}{\sum_{G_F,G_M,G_C}P(D_F,D_M,D_C|G_F,G_M,G_C)P(G_F,G_M,G_C)}
\end{equation}


\begin{align}\label{eqn:model}
P(D_F,D_M,D_C|G_F,G_M,G_C) &= P(D_F|G_F)P(D_M|G_M)P(D_C|G_M) \\
	&= \left(\prod_{f=1}^{|D_F|}P(R_f|G_F)\right)\left(\prod_{m=1}^{|D_M|}P(R_m|G_M)\right)\left(\prod_{c=1}^{|D_C|}P(R_c|G_C)\right)
\end{align}

\begin{equation}\label{eqn:model}
P(G_F,G_M,G_C) = P(G_C|G_F,G_M)P(G_F)P(G_M)
\end{equation}

\[ P(R|G) = \left\{ 
  \begin{array}{l l}
    1-\epsilon_{seq} & \quad \text{if $G$ is homozygous and $R \in G$}\\
    \epsilon_{seq}/3 & \quad \text{if $G$ is homozygous and $R \notin G$}\\    
      0.5(1-2\epsilon_{seq}/3) & \quad \text{if $G$ is heterozygous and $R \in G$}\\
    \epsilon_{seq}/3 & \quad \text{if $G$ is heterozygous and $R \notin G$}
  \end{array} \right.\]
  

\[ P(G_C|G_F,G_M) = \left\{ 
  \begin{array}{l l}
    \epsilon_{denovo}/N_{denovo} & \quad \text{if $G_C$ is a denovo mutation}\\
    (1-\epsilon_{denovo})/N_{mendelian} & \quad \text{otherwise}
  \end{array} \right.\]

\begin{align*}
L_{denovo} &= \sum_{G_F,G_M,G_C}P(D_F,D_M,D_C|G_F,G_M,G_C)P(G_F,G_M,G_C)I_{denovo}(G_F,G_M,G_C)\\
L_{mendelian} &= \sum_{G_F,G_M,G_C}P(D_F,D_M,D_C|G_F,G_M,G_C)P(G_F,G_M,G_C)I_{mendelian}(G_F,G_M,G_C)
\end{align*}

\begin{align}
\frac{L_{denovo}}{L_{denovo} + L_{mendelian}} > 0.5
\end{align}

\begin{align}
\frac{L_{denovo}}{L_{mendelian}} > \mathcal{T}
\end{align}

\begin{align}
I_{denovo}(\argmax_{G_F,G_M,G_C} P(G_F,G_M,G_C|D_F,D_M,D_C))
\end{align}


\begin{figure}
\centering
\begin{tikzpicture}[->,>=stealth',shorten >=1pt,auto,node distance=2.8cm,
                    semithick]
  \tikzstyle{every state}=[fill=red,draw=none,text=white]
  \tikzstyle{parent}=[circle,thick,draw=blue!75,fill=blue!20,minimum size=6mm]
  \tikzstyle{child}=[parent,draw=red!75,fill=red!20]
  \tikzstyle{data}=[rectangle,thick,draw=black!75,
  			  fill=black!20,minimum size=4mm]

  \node[child] 		   (C)                    {$G_C$};
  \node[parent]         (F) [above left of=C] {$G_F$};
  \node[parent]         (M) [above right of=C] {$G_M$};
  \node[data]         (DC) [below of=C] 		{$D_M$};
  \node[data]         (DF) [below of=F]       {$D_F$};
  \node[data]         (DM) [below of=M]       {$D_C$};

  \path (F) edge              node {} (C)
            edge              node {} (DF)
        (M) edge              node {} (C)
            edge              node {} (DM)
        (C) edge              node {} (DC);
        
\end{tikzpicture}
\caption{The hidden variables $G_{F}, G_{M} and G_{C}$ correspond to the genotypes of the father, mother and child respectively. The genotypes take one among 10 values from $AA, AC, AT, AG, CC, CT, CG, GG, GT$ and $TT$. The trio of genotypes values can thus have $10^3$ possible values. The observed variables ($D_F, D_M$ and $D_C$) are the bases present at the candidate position for a particular read }
\label{fig:bayesnet}
\end{figure}

\subsection{Algorithm}

\section{Design}




\begin{align}\label{eqn:diffeq}
\frac{dC(t)}{dt}  &= \sum_{i \in \mathcal{V}}\alpha_i V_i(t) - \sum_{j \in \mathcal{P}}\beta_j C(t)\\
\frac{dP_j(t)}{dt}  &= \beta_j C(t) - \sum_{i \in \mathcal{V}}\gamma_{ij}V_j(t) \\
\frac{dV_i(t)}{dt}  &= \sum_{j \in \mathcal{P}}\gamma_{ij}P_j(t) - \alpha_i V_i(t)
\end{align}

\begin{figure}
\centering
\begin{tikzpicture}[->,>=stealth',shorten >=1pt,auto,node distance=3cm,thick,
  pleasure node/.style={circle,fill=red!20,draw,font=\sffamily\bfseries},
  blank node/.style={circle,fill=blue!20,fill opacity=0,font=\sffamily\bfseries},
  bucket/.style={draw, fill=blue!20, minimum height=3em, minimum width=4em, cylinder, shape border rotate=90, shape aspect=0.1}]


    \node [bucket] (Vw)                                    {$V_{work}(t)$};
    \node [bucket] (Vf) [below of=Vw]                      {$V_{family}(t)$};

    \node [bucket, minimum height=5em, minimum width=4em, fill=green!20] (C) [right of=Vw, yshift = -15mm]                         {$C(t)$};


    \node [pleasure node] (Pj) [right of=C]                      {$P_{joy}(t)$};
    \node [pleasure node] (Pp) [above of=Pj]                       {$P_{peace}(t)$};
    \node [pleasure node] (Pt)  [below of=Pj] 					 {$P_{thrill}(t)$};

 \path[every node/.style={font=\sffamily\small}]
    (Vw) edge node [below] {$\alpha_w$} (C)
    (Vf) edge node [above]   {$\alpha_f$} (C)

    (C) edge  node [below] {$\beta_p$} (Pp)
     edge node [above]   {$\beta_j$} (Pj)
     edge  node [above] {$\beta_t$} (Pt)

	(Pp) edge [bend right] node [above] {$\gamma_{pw}$} (Vw)
	 edge [bend right] node [above] {$\gamma_{pf}$} (Vf)
	 
	(Pj) edge [bend right] node [above] {$\gamma_{jw}$} (Vw)
	 edge [bend left] node [below] {$\gamma_{jf}$} (Vf)
	 
	(Pt) edge [bend left] node [below] {$\gamma_{tw}$} (Vw)
	 edge [bend left] node [below] {$\gamma_{tf}$} (Vf);

  \begin{pgfonlayer}{background}
    \filldraw [line width=4mm,join=round,black!10]
      (Vw.north  -| Vw.east)  rectangle (Vf.south  -| Vf.west)
      (C.north  -| C.east)  rectangle (C.south  -| C.west)
      (Pp.north  -| Pp.east)  rectangle (Pt.south  -| Pt.west);
  \end{pgfonlayer}
\end{tikzpicture}
\caption{A simple instance of Bucket Theory. There are two value buckets $\mathcal{V}=\{work,family\}$, the control bank $C$ and three nutrients at the pleasure Caf\'e $\mathcal{P}=\{peace, joy, thrill\}$.  The value resources are converted into a control currency, $C(t)$. Control is then used to purchase the different pleasure nutrients $P_j(t) ; j\in \mathcal{P}$. }
\label{fig:pleasure}
\end{figure}

\subsection{Commandline usage}


Some citations are ~\cite{Li2012} and~\cite{Michaelson2012}.

\section{Experiments and Validation}

\subsection{NA12878 trio}

\subsection{Benchmarking}

\section{Conclusion}

\begin{thebibliography}{10}

\bibitem{Li2012}
   Li B et. al,
   \emph{A likelihood-based framework for variant calling and de novo mutation detection in families},
   PLoS Genetics, 2012
   Volume 8, Number 10, Pages e1002944

\bibitem{Michaelson2012}   
   Michaelson et al,
   \emph{Whole-Genome Sequencing in Autism Identifies Hot Spots for De Novo Germline Mutation},  
   Cell, 2012, 
   Volume 151, Number 7, Pages 1431 - 1442

\bibitem{Jordan2008}
	Wainwright, Jordan,
	\emph{Graphical Models, Exponential Families, and Variational Inference},
	Foundations and Trends in Machine Learning, 2008,
	Volume 1, Issue 1-2, pp 1-305	
	
\end{thebibliography}
\end{document}